src/basic_recipes/datashader.jl

```diff
@@ -0,0 +1,388 @@
+        if isfinite(final_value)
+            mini = min(final_value, mini)
+            maxi = max(final_value, maxi)
+        end
+        return final_value
+    end
+    return (mini, maxi)
+end
+
+function aggregation_implementation!(
+        ::AggThreads,
+        aggbuffer::AbstractVector, pixelbuffer::AbstractVector,
+        c::Canvas, op::AggOp,
+        points, local_op, point_func)
+    xmin, xmax = xlims(c)
+    ymin, ymax = ylims(c)
+    xsize, ysize = size(c)
+    xmax > xmin || error("require xmax > xmin")
+    ymax > ymin || error("require ymax > ymin")
+    # by adding eps to width we can use the scaling factor plus floor directly to compute the bin indices
+    xwidth = xmax - xmin
+    xscale = xsize / (xwidth + eps(xwidth))
+    ywidth = ymax - ymin
+    yscale = ysize / (ywidth + eps(ywidth))
+
+    # each thread reduces some of the data separately
+    @assert length(aggbuffer) == Threads.nthreads() * xsize * ysize
+    @assert length(pixelbuffer) == xsize * ysize
+    @assert eltype(aggbuffer) === typeof(null(op))
+
+    # using ReshapedArray directly like this is not advised, but as it lives only briefly it should be ok
+    out = Base.ReshapedArray(aggbuffer, (xsize, ysize, Threads.nthreads()), ())
+    out2 = Base.ReshapedArray(pixelbuffer, (xsize, ysize), ())
+    n = length(points)
+    chunks = round.(Int, range(1, n, length = Threads.nthreads()+1))
+
+    @threads for t in 1:Threads.nthreads()
+        from = chunks[t]
+        to = chunks[t+1]
+        for idx in from:to
+            p = @inbounds point_func(points[idx])
+            x = p[1]
+            y = p[2]
+            if length(p) > 2 # should compile away
+                z = p[3]
+            end
+            xmin ≤ x ≤ xmax || continue
+            ymin ≤ y ≤ ymax || continue
+            i = 1 + floor(Int, xscale*(x-xmin))
+            j = 1 + floor(Int, yscale*(y-ymin))
+            if length(p) == 2 # should compile away
+                @inbounds out[i,j,t] = update(op, out[i,j,t])
+            elseif length(p) == 3
+                @inbounds out[i,j,t] = update(op, out[i,j,t], z)
+            end
+        end
+    end
+    # reduce along the thread dimension
+    mini, maxi = Inf, -Inf
+    for j in 1:ysize
+        for i in 1:xsize
+            @inbounds val = out[i,j,1]
+            for t in 2:Threads.nthreads()
+                @inbounds val = merge(op, val, out[i,j,t])
+            end
+            # update the value in out2 directly in this loop
+            final_value = local_op(value(op, val))
+            if isfinite(final_value)
+                mini = min(final_value, mini)
+                maxi = max(final_value, maxi)
+            end
+            @inbounds out2[i, j] = final_value
+        end
+    end
+    return (mini, maxi)
+end
+
+
+export AggAny, AggCount, AggMean, AggSum, AggSerial, AggThreads
+
+end
+
+using ..PixelAggregation
+
+function equalize_histogram(matrix; nbins=256 * 256)
```

Comment:
I'm not sure why `nbins=256 * 256` was chosen here. Doesn't each bin get one step in the final colormap? I think `nbins = 256` might be more appropriate then, as in 8bit color depth as is usually available.